### ISSUE TITLE
freshen top-level names?

### ISSUE THREAD

**Issue by stchang** [vishesh/racketscript#178]
I forget the reason why the freshen pass does not include top-level names. @vishesh do you remember?

Because Racket programs could expand to multiple functions with the same name (disambiguated with scopes).

Right now it's causing the tests to fail because `sort` infinite loops (see https://github.com/racket/racket/issues/3471), but other cases could easily run into this problem as well.

**Comment by vishesh** [comment]
I don't recall any specific reason, but I think it may not be technical, just to have reasonable names when using Racket modules from JS side (interop) initially. 

I may be wrong here, but IIRC we don't compile function in other compilation level, so doing `(for-require -X)` to bring that module's level-1 into current level 0 wouldn't work, and I had plans to fix that. Maybe, I think the issue of handling same name different scope was probably planned to handled there.

**Comment by stchang** [comment]
Yah, while trying to address the issue, I understand now that it's probably infeasible to try to freshen things from `racket` core.

I think I have a workaround that should only freshen internal unexported names.

**Comment by vishesh** [comment]
Yeah, I think that should work. I think only one of them can be exported?

```
#lang racket/base

(define-syntax-rule (d)
  (begin
    (define example 1)
    (provide example)))

(d)
(d)
```

I think one easy way would be freshen all top-level names and make no exception there. And when we are looking at `provide` forms, we can generate something like `export sort918234 as sort`. Not sure if that would create problem for this test `https://github.com/vishesh/racketscript/blob/master/tests/modules/macro-intrduced-binding.rkt`.
